File: 97. Interleaving String.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        int m = s1.size(), n = s2.size();
        if (m + n != s3.size()) return false;
        
        vector<vector<bool> > dp(m + 1);
        for (int i = 0; i < dp.size(); i++) 
            dp[i].resize(n + 1);
        
        dp[0][0] = true;
        
        // 先考虑分别单独用S1或者S2能否构成S3
        for(int i = 1; i < m+1 && s1[i-1] == s3[i-1]; i++)
            dp[i][0] = true;
        for(int i = 1; i < n+1 && s2[i-1] == s3[i-1]; i++)
            dp[0][i] = true;
        
        for (int i = 1; i < m+1; i++) {
            for (int j = 1; j < n+1; j++) {
                if((s3[i+j-1] == s1[i-1]&&dp[i-1][j]) || 
                    (s3[i+j-1] == s2[j-1]&&dp[i][j-1]))
                    dp[i][j] = true;
            }
        }
        
        return dp[m][n];
    }
};

File: python/97. Interleaving String.py
class Solution(object):
    def isInterleave(self, s1, s2, s3):
        """
        :type s1: str
        :type s2: str
        :type s3: str
        :rtype: bool
        """
        n1, n2, n3 = len(s1), len(s2), len(s3)
        if n1+n2!=n3: return False
        dp = [[False]*(n2+1) for _ in range(n1+1)]
        dp[0][0] = True
        for i in range(1, n1+1):
            dp[i][0] = dp[i-1][0] and s1[i-1]==s3[i-1]
        for j in range(1, n2+1):
            dp[0][j] = dp[0][j-1] and s2[j-1]==s3[j-1]
        
        for i in range(1, n1+1):
            for j in range(1, n2+1):
                dp[i][j] = (dp[i-1][j] and s1[i-1]==s3[i-1+j] ) or (dp[i][j-1] and s2[j-1]==s3[j-1+i])
        
        return dp[n1][n2]

# 思路：dp[i][j]表示s1的前i-1个字符，s2的前j-1个字符能够构成s3的前i+j-1个字符,
# dp[i][j] = (dp[i-1][j] and s1[i-1]==s3[i-1+j] ) or (dp[i][j-1] and s2[j-1]==s3[j-1+i])
